crates/ide/src/prime_caches.rs

```diff
@@ -3,10 +3,65 @@
 //! request takes longer to compute. This modules implemented prepopulating of
 //! various caches, it's not really advanced at the moment.
 
-use crate::{FileId, RootDatabase};
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
 
-pub(crate) fn prime_caches(db: &RootDatabase, files: Vec<FileId>) {
-    for file in files {
-        let _ = crate::syntax_highlighting::highlight(db, file, None, false);
-    }
+use base_db::salsa::ParallelDatabase;
+use base_db::SourceDatabase;
+use hir::db::DefDatabase;
+
+use crate::RootDatabase;
+
+#[derive(Debug)]
+pub enum PrimeCachesProgress {
+    Started,
+    /// We started indexing a crate.
+    StartedOnCrate {
+        on_crate: String,
+        n_done: usize,
+        n_total: usize,
+    },
+    /// We finished indexing all crates.
+    Finished,
+}
+
+pub(crate) fn prime_caches<F>(db: &RootDatabase, cb: F)
+where
+    F: Fn(PrimeCachesProgress) + Sync,
+{
+    let _p = profile::span("prime_caches");
+    let graph = db.crate_graph();
+    let topo = &graph.crates_in_topological_order();
+
+    let cb = &cb;
+    let finished = &AtomicUsize::new(0);
+    let db = db.snapshot();
+
+    cb(PrimeCachesProgress::Started);
+
+    rayon::scope(move |s| {
+        for krate in topo {
+            let crate_name = graph[*krate]
+                .declaration_name
+                .as_ref()
+                .map(ToString::to_string)
+                .unwrap_or_default();
+
+            let db = db.snapshot();
+            s.spawn(move |_| {
+                cb(PrimeCachesProgress::StartedOnCrate {
+                    on_crate: crate_name,
+                    n_done: finished.load(Ordering::Relaxed),
+                    n_total: topo.len(),
+                });
+                db.crate_def_map(*krate);
+                let old = finished.fetch_add(1, Ordering::Relaxed);
+                assert!(old < topo.len());
+            });
+        }
+    });
+
+    cb(PrimeCachesProgress::Finished);
+
+    assert_eq!(finished.load(Ordering::Relaxed), topo.len());
```

Comment:
Could declare `finished` as `mut` and save an atomic load here: https://doc.rust-lang.org/stable/std/sync/atomic/struct.AtomicUsize.html#method.get_mut! Probably doesn't worth it ;)

Reply:
`into_inner` would also work